File: 2017/09_StreamProcessing/groups.py
# ======================================================================
# Stream Processing
#   Advent of Code 2017 Day 09 -- Eric Wastl -- https://adventofcode.com
#
# Computer simulation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                       g r o u p s . p y
# ======================================================================
"A solver for Stream Processing for Advent of Code 2017 Day 09"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------

from collections import namedtuple

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------

StateTableRow = namedtuple('StateTableRow',
                           'state, inchr, nxtst, delta, trash')

STATE_TABLE = [
    StateTableRow(0, '{', 1, 1, 0),
    StateTableRow(1, '{', 1, 1, 0),
    StateTableRow(1, '}', 1, -1, 0),
    StateTableRow(1, '<', 2, 0, 0),
    StateTableRow(2, '>', 1, 0, 0),
    StateTableRow(2, '!', 3, 0, 0),
    StateTableRow(2, None, 2, 0, 1),  # To count garbage characters
    StateTableRow(3, None, 2, 0, 0)
]

# ======================================================================
#                                                                 Groups
# ======================================================================


class Groups(object):
    """Object processing group / garbage stream"""

    def __init__(self, part2=False, text=None, verbose=False, limit=0):

        # 1. Set the initial values
        self.part2 = part2
        self.state = 0
        self.level = 0
        self.group = 0
        self.score = 0
        self.trash = 0

        # 2. If text is not None, process it
        if text is not None:
            self.process_stream(text=text, verbose=verbose, limit=limit)

    def __str__(self):
        return "s=%d l=%d g=%d s=%d" % (
            self.state, self.level, self.group, self.score)


    def process_stream(self, text=None, verbose=False, limit=0):
        "Process the text of a stream"

        # 1. If no text, not much to do
        if text is None:
            return

        # 2. Loop for all of the character
        for cnum, char in enumerate(text):

            # 3. Process the character
            self.process_character(char, verbose=verbose)

            # 4. Check if processing limit reached
            if cnum > limit > 0:
                break

    def process_character(self, char, verbose=False):
        "Process the text of a stream"

        # 1. Loop for the all of the rows in the state table
        for row in STATE_TABLE:

            # 2. Ignore row if state does not match
            if self.state != row.state:
                continue

            # 3. Ignore row if character does not match
            if row.inchr is not None and row.inchr != char:
                continue

            # 4. Set up adjustments
            next_state = row.nxtst
            next_level = self.level + row.delta
            if row.delta == -1:
                self.group += 1
                self.score += self.level
            self.trash += row.trash

            # 5. Describe the state (if desired)
            if verbose:
                print("state %d, char %s -> next %d, level %d, group %d score %d" %
                      (self.state, char, next_state, next_level, self.group, self.score))

            # 6. Set the next state and level
            self.state = next_state
            self.level = next_level

            # 7. Character processed, exit loop
            break

# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                         g r o u p s . p y                      end
# ======================================================================

File: 2017/09_StreamProcessing/groups.lua
-- ======================================================================
-- StreamProcessing
--   Advent of Code 2017 Day 09 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                         g r o u p s . l u a
-- ======================================================================
-- A solver for the Advent of Code 2017 Day 09 puzzle

-- ----------------------------------------------------------------------
--                                                                  local
-- ----------------------------------------------------------------------
local Groups = { part2 = false, text = {} }

-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------
local state_table = {
  [1] = {["{"] = 7, ["<"] = 9, [">"] = 9, ["}"] = 9, [","] = 9, ["!"] = 9, ["?"] = 9 }, -- 1 Initial
  [2] = {["{"] = 7, ["<"] = 3, [">"] = 9, ["}"] = 8, [","] = 2, ["!"] = 9, ["?"] = 9 }, -- 2 In Brace
  [3] = {["{"] = 6, ["<"] = 6, [">"] = 2, ["}"] = 6, [","] = 6, ["!"] = 4, ["?"] = 6 }, -- 3 In Garbage
  [4] = {["{"] = 3, ["<"] = 3, [">"] = 3, ["}"] = 3, [","] = 3, ["!"] = 3, ["?"] = 3 }, -- 4 In Shriek
}  
  
-- ======================================================================
--                                                                 Groups
-- ======================================================================

-- Object for StreamProcessing

function Groups:Groups(o)

  -- 1. Set the initial values
  o = o or {}
  o.part2 = o.part2 or false
  o.text = o.text or {}

  -- 2. Create the object metatable
  self.__index = self
  setmetatable(o, self)

  return o
end

function Groups:next_state(state, character)
  -- Return the next state
  
  -- 1. Get the current row in the state table
  local row = state_table[state]
  
  -- 2. Get the next state
  local next_state = row[character]
  if next_state == nil then
    next_state = row["?"]
  end
  
  -- 3. Return the next state
  return next_state
end

function Groups:score_braces(line)
  -- Score the braces
  
  -- 1. Initial state
  local score = 0
  local state = 1
  local depth = 0
  
  -- 2. Loop for all characters of the string
  for indx = 1, #line do
    local character = line:sub(indx, indx)
    
    -- 3. Get the next state
    local new_state = self:next_state(state, character)
    -- print("score_braces", state, character, new_state, state, depth, indx)
    
    -- 4. Process special states
    if new_state == 9 then
      print("Unexpected character", character, score, state, depth, indx)
      return -1
    elseif new_state == 8 then
      score = score + depth
      depth = depth - 1
      if depth == 0 then
        new_state = 1
      else
        new_state = 2
      end
    elseif new_state == 7 then
      depth = depth + 1
      new_state = 2
    elseif new_state == 6 then
      new_state = 3
    end
    
    -- 5. Set the state
    state = new_state
  end
  
  -- 6. Return the score
  return score
end

function Groups:score_garbage(line)
  -- Score the braces
  
  -- 1. Initial state
  local score = 0
  local state = 1
  local depth = 0
  
  -- 2. Loop for all characters of the string
  for indx = 1, #line do
    local character = line:sub(indx, indx)
    
    -- 3. Get the next state
    local new_state = self:next_state(state, character)
    -- print("score_braces", state, character, new_state, state, depth, indx)
    
    -- 4. Process special states
    if new_state == 9 then
      print("Unexpected character", character, score, state, depth, indx)
      return -1
    elseif new_state == 8 then
      depth = depth - 1
      if depth == 0 then
        new_state = 1
      else
        new_state = 2
      end
    elseif new_state == 7 then
      depth = depth + 1
      new_state = 2
    elseif new_state == 6 then
      score = score + 1
      new_state = 3
    end
    
    -- 5. Set the state
    state = new_state
  end
  
  -- 6. Return the score
  return score
end
  
function Groups:part_one(args)
  -- Returns the solution for part one

  -- 0. Precondition axioms
  local verbose = args.verbose or false
  local limit = args.limit or 0
  assert(verbose == true or verbose == false)
  assert(limit >= 0)

  -- 1. Return the solution for part one
  return self:score_braces(self.text[1])
end


function Groups:part_two(args)
  -- Returns the solution for part two

  -- 0. Precondition axioms
  local verbose = args.verbose or false
  local limit = args.limit or 0
  assert(verbose == true or verbose == false)
  assert(limit >= 0)

  -- 1. Return the solution for part two
  return self:score_garbage(self.text[1])
end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------

return Groups

-- ======================================================================
-- end                       g r o u p s . l u a                      end
-- ======================================================================
